include/swift/AST/Attr.h

```diff
@@ -1918,17 +1906,20 @@ class DerivativeAttr final
                                 SourceLoc atLoc, SourceRange baseRange,
                                 TypeRepr *baseTypeRepr,
                                 DeclNameRefWithLoc original,
-                                IndexSubset *parameterIndices);
+                                IndexSubset *parameterIndices,
+                                LazyMemberLoader *resolver, uint64_t data);
 
   TypeRepr *getBaseTypeRepr() const { return BaseTypeRepr; }
   DeclNameRefWithLoc getOriginalFunctionName() const {
     return OriginalFunctionName;
   }
-  AbstractFunctionDecl *getOriginalFunction(ASTContext &context) const;
-  void setOriginalFunction(AbstractFunctionDecl *decl);
-  void setOriginalFunctionResolver(LazyMemberLoader *resolver,
-                                   uint64_t resolverContextData);
 
+  /// Retrieves the referenced original function declaration. Returns \c nullptr
+  /// and emits diagnostics on error.
+  AbstractFunctionDecl *
+  getOriginalFunction(const AbstractFunctionDecl *derivative) const;
```

Comment:
Note: `DeclAttribute`s don’t store the declaration on which they’re declared, so attribute-related requests ~always take that declaration as an argument. This makes attribute “getter”/"setter" functions a bit cumbersome, e.g. `DerivativeAttr::getOriginalFunction(const AbstractFunctionDecl *derivative)` and `TypeEraserAttr::getResolvedType(const ProtocolDecl *PD)`.

I wonder if Is this intentional? Or is there a way to avoid the declaration argument to improve attribute “getter”/"setter" usability?

Reply:
That's an inversion of responsibility that we should avoid. If only abstract functions can have this attribute applied (is that true? Can I mark a stored property closure's derivative?) then the accessor that calls the request should appear on `AbstractFunctionDecl`.

Reply:
> That's an inversion of responsibility that we should avoid. If only abstract functions can have this attribute applied (is that true? Can I mark a stored property closure's derivative?) then the accessor that calls the request should appear on `AbstractFunctionDecl`.

Yes, actually only `FuncDecl` can have `@derivative` attributes. I can further constrain types.
Yes, I saw that `ValueDecl *ValueDecl::getDynamicallyReplacedDecl()` is designed that way.

However, multiple `@derivative` attributes are currently allowed on a declaration, so `AbstractFunctionDecl *FuncDecl::getDerivativeAttrOriginalDeclaration()` wouldn't work.

I can try adding `AbstractFunctionDecl *FuncDecl::getDerivativeAttrOriginalDeclaration(const DerivativeAttr *attr)`, if that makes sense.

Reply:
Grabbing all of the attached derivative attribute information is a nice approach. If we can do just the resolved differentiation information that would ultimately be a cleaner solution.